File: src/torchcodec/_core/AudioConverter.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

#include "AudioConverter.h"

#include <vector>

#include "AudioCommon.h"

namespace facebook::torchcodec {

AudioConverter::AudioConverter(
    std::optional<int> sample_rate,
    std::optional<int> num_channels)
    : requested_sample_rate_(sample_rate),
      requested_num_channels_(num_channels) {
  STD_TORCH_CHECK(
      !sample_rate.has_value() || *sample_rate > 0,
      "sample_rate must be > 0. Got: ",
      sample_rate.value_or(0));
  STD_TORCH_CHECK(
      !num_channels.has_value() || *num_channels > 0,
      "num_channels must be > 0. Got: ",
      num_channels.value_or(0));
}

void AudioConverter::reset() {
  swr_context_.reset();
  src_sample_format_ = AV_SAMPLE_FMT_NONE;
  src_sample_rate_ = 0;
  src_num_channels_ = 0;
  out_sample_rate_ = 0;
  out_num_channels_ = 0;
}

torch::stable::Tensor AudioConverter::convert(
    const torch::stable::Tensor& samples,
    int sample_rate) {
  STD_TORCH_CHECK(
      samples.dim() == 2,
      "Expected a 2D [num_channels, num_samples] tensor, got a ",
      samples.dim(),
      "D one.");
  STD_TORCH_CHECK(samples.is_contiguous(), "The samples must be contiguous.");
  STD_TORCH_CHECK(sample_rate > 0, "sample_rate must be > 0.");

  AVSampleFormat src_sample_format =
      planar_sample_format(samples.scalar_type());
  int num_channels = static_cast<int>(samples.sizes()[0]);
  int num_samples = static_cast<int>(samples.sizes()[1]);
  STD_TORCH_CHECK(
      num_channels > 0, "The samples must have at least 1 channel.");

  if (swr_context_ == nullptr) {
    src_sample_format_ = src_sample_format;
    src_sample_rate_ = sample_rate;
    src_num_channels_ = num_channels;
    out_sample_rate_ = requested_sample_rate_.value_or(sample_rate);
    out_num_channels_ = requested_num_channels_.value_or(num_channels);
    swr_context_.reset(create_swr_context(
        src_sample_format_,
        kAudioOutSampleFormat,
        src_sample_rate_,
        out_sample_rate_,
        src_num_channels_,
        out_num_channels_));
  } else {
    // swresample is configured once, from the first samples we see, and its
    // buffered state is tied to that configuration. Rather than silently
    // reconfiguring - which would discard whatever it still holds - we make the
    // caller decide, by reset()ing.
    STD_TORCH_CHECK(
        src_sample_format == src_sample_format_ &&
            sample_rate == src_sample_rate_ &&
            num_channels == src_num_channels_,
        "This AudioConverter was set up for ",
        src_num_channels_,
        " channels of ",
        av_get_sample_fmt_name(src_sample_format_),
        " at ",
        src_sample_rate_,
        " Hz, but got ",
        num_channels,
        " channels of ",
        av_get_sample_fmt_name(src_sample_format),
        " at ",
        sample_rate,
        " Hz. Call reset() to convert a different stream.");
  }

  const auto* base = static_cast<const uint8_t*>(samples.const_data_ptr());
  int64_t bytes_per_channel =
      num_samples * av_get_bytes_per_sample(src_sample_format);
  std::vector<const uint8_t*> src_planes(num_channels);
  for (int channel = 0; channel < num_channels; ++channel) {
    src_planes[channel] = base + channel * bytes_per_channel;
  }

  return swr_convert_to_tensor(
      swr_context_,
      src_planes.data(),
      num_samples,
      out_num_channels_,
      get_swr_output_num_samples_bound(
          swr_context_, num_samples, src_sample_rate_, out_sample_rate_));
}

torch::stable::Tensor AudioConverter::drain() {
  STD_TORCH_CHECK(
      swr_context_ != nullptr,
      "This AudioConverter hasn't converted any samples, so there is nothing "
      "to drain and no way to know what shape the result should have.");
  // A null input is what tells swr_convert() to flush. Unlike the convert()
  // path we ask swresample how much it is holding rather than deriving a bound
  // from an input size, since here there is no input.
  return swr_convert_to_tensor(
      swr_context_,
      /*src_planes=*/nullptr,
      /*num_src_samples=*/0,
      out_num_channels_,
      swr_get_out_samples(swr_context_.get(), 0));
}

} // namespace facebook::torchcodec

File: src/torchcodec/decoders/_blocks/_audio_converter.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

from __future__ import annotations

from torchcodec._core.ops import (
    _blocks_audio_converter_convert,
    _blocks_audio_converter_drain,
    _blocks_audio_converter_reset,
    _blocks_create_audio_converter,
)
from torchcodec._frame import AudioSamples

from ._frame import RawAudioSamples
from ._helpers import _process_local


@_process_local(
    "A resampler carries filter history across calls that FFmpeg cannot "
    "serialise. Construct one in each process."
)
class AudioConverter:
    """Turn :class:`RawAudioSamples` into normalised float32
    :class:`~torchcodec.AudioSamples`, optionally resampling and remixing
    channels.

    Low-level API: for straightforward decoding, use
    :class:`~torchcodec.decoders.AudioDecoder` instead.

    .. code-block:: python

        converter = AudioConverter(sample_rate=16_000)

        for packet in demuxer:
            for raw_samples in packet_decoder.decode(packet):
                samples = converter.convert(raw_samples)
        for raw_samples in packet_decoder.drain():
            samples = converter.convert(raw_samples)
        samples = converter.drain()

    Unlike a :class:`ColorConverter`, this object is a stateful stream
    processor, and it is bound to an audio stream: feed it the stream's samples
    in order. When resampling, it holds samples back between calls, so you won't
    necessarily get the same number of samples out as you put in for a given
    call to :meth:`convert`.

    Args:
        sample_rate (int, optional): The output sample rate. Defaults to the
            source's own, i.e. no resampling.
        num_channels (int, optional): The output number of channels. Defaults
            to the source's own.
    """

    def __init__(self, sample_rate: int | None = None, num_channels: int | None = None):
        self._handle = _blocks_create_audio_converter(
            sample_rate=sample_rate, num_channels=num_channels
        )
        self._requested_sample_rate = sample_rate
        self._drained = False

        self._first_frame_pts_seconds: float | None = None
        self._out_sample_rate: int | None = None
        self._num_emitted_samples = 0
        # The demuxer position these samples come from. See Packet._generation:
        # a resampler carries state across calls, so a seek invalidates it just
        # as it invalidates the decoder's.
        self._generation: int | None = None

    def _wrap(self, data) -> AudioSamples:
        assert self._out_sample_rate is not None  # mypy
        assert self._first_frame_pts_seconds is not None  # mypy
        # We manually compute the pts of all frames but the first one: when
        # resampling happens, libswresample buffers samples and emits them
        # later. Not all the samples of the first frame are necessarily emitted
        # immediately, they may be emitted with the next frame. Without this,
        # we'd fail the test_audio_converter_pts_is_contiguous() test.
        pts_seconds = (
            self._first_frame_pts_seconds
            + self._num_emitted_samples / self._out_sample_rate
        )
        self._num_emitted_samples += data.shape[1]
        return AudioSamples(
            data=data,
            pts_seconds=pts_seconds,
            duration_seconds=data.shape[1] / self._out_sample_rate,
            sample_rate=self._out_sample_rate,
        )

    def convert(self, raw_samples: RawAudioSamples) -> AudioSamples:
        """Convert one :class:`RawAudioSamples` into normalised float32
        :class:`~torchcodec.AudioSamples`.

        You may not get the same number of samples out as you put in, especially
        if you are resampling.

        Returns:
            The converted samples, normalised float32 in ``[-1, 1]``. When
            resampling, fewer than were passed in, possibly none.
        """
        if self._generation is None:
            self._generation = raw_samples._generation
        elif self._generation != raw_samples._generation:
            raise RuntimeError(
                "The demuxer seeked since this converter was last reset(), and "
                "a resampler carries state across calls, so these samples "
                "would be resampled against the wrong history. Call reset() on "
                "every converter fed by that demuxer after a seek."
            )
        if self._drained:
            raise RuntimeError(
                "This AudioConverter has been drained. Call reset() to convert "
                "more samples."
            )
        if self._first_frame_pts_seconds is None:
            self._first_frame_pts_seconds = raw_samples.pts_seconds
            self._out_sample_rate = (
                self._requested_sample_rate
                if self._requested_sample_rate is not None
                else raw_samples.sample_rate
            )
        data = _blocks_audio_converter_convert(
            self._handle, raw_samples.data, raw_samples.sample_rate
        )
        return self._wrap(data)

    def drain(self) -> AudioSamples:
        """Return the samples the resampler was still holding on to.

        Empty unless you are resampling. This is not the codec's own buffer,
        which :meth:`AudioPacketDecoder.drain` takes care of.
        """
        if self._first_frame_pts_seconds is None:
            raise RuntimeError(
                "This AudioConverter hasn't converted any samples, so there is "
                "nothing to drain."
            )
        data = _blocks_audio_converter_drain(self._handle)
        self._drained = True
        return self._wrap(data)

    def reset(self) -> None:
        """Drop the resampler's state and start over.

        Needed after a :meth:`Demuxer.seek`, after :meth:`drain`, and before
        converting a different stream.
        """
        _blocks_audio_converter_reset(self._handle)
        self._drained = False
        self._first_frame_pts_seconds = None
        self._out_sample_rate = None
        self._generation = None
        self._num_emitted_samples = 0
